cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,347 @@
+import numpy as np
+from typing import List, Set, Tuple, Sequence, Dict, Any
+
+from matplotlib import pyplot as plt
+from cirq import devices, ops, circuits, sim
+
+
+class CrossEntropyResult:
+    """Results from a cross-entropy benchmarking (XEB) experiment."""
+
+    def __init__(self, num_cycle_range: Sequence[int],
+                 xeb_fidelities: Sequence[float]):
+        """
+        Args:
+            num_cycle_range: The different numbers of cycles (circuit depth) at
+                which the measurements were taken.
+            xeb_fidelities: The average XEB fidelity at each number of cycles.
+        """
+        self._num_cycles = num_cycle_range
+        self._xeb_fidelities = xeb_fidelities
+
+    @property
+    def data(self) -> Sequence[Tuple[int, float]]:
+        """Returns a sequence of tuple pairs with the first item being a
+        number of cycles and the second item being the corresponding average
+        XEB fidelity.
+        """
+        return [(num, prob) for num, prob in zip(self._num_cycles,
+                                                 self._xeb_fidelities)]
+
+    def plot(self, **plot_kwargs: Any) -> None:
+        """Plots the average XEB fidelity vs the number of cycles.
+
+        Args:
+            **plot_kwargs: Arguments to be passed to matplotlib.pyplot.plot.
+        """
+        fig = plt.figure()
+        ax = plt.gca()
+        ax.set_ylim([0, 1.1])
+
+        plt.plot(self._num_cycles, self._xeb_fidelities, 'ro-',
+                 figure=fig, **plot_kwargs)
+        plt.xlabel('Number of Cycles', figure=fig)
+        plt.ylabel('XEB Fidelity', figure=fig)
+        fig.show()
+
+
+def cross_entropy_benchmarking(
+        sampler: sim.Sampler,
+        qubits: Sequence[devices.GridQubit],
+        two_qubit_gate: ops.gate_features.TwoQubitGate,
+        *,
+        num_circuits: int = 20,
+        repetitions: int = 1000,
+        num_cycle_range: Sequence[int] = range(2, 103, 10),
+        interaction_sequence: List[Set[Tuple[devices.GridQubit,
```

Comment:
This is a pretty cumbersome argument to specify. I'm not sure how to make it better.

The way you separate out the two qubit gate is perhaps not necessary. This argument could be a `List[Set[Operation]]` and it would work just as well but be more flexible.

Reply:
This is now modified as suggested. The user would specify a list of sets of operations to be interleaved with the single-qubit gates. They can also specify nothing, in which case the circuits would just be single-qubit gates.